### PR TITLE
Ee prefect

### PR DESCRIPTION
**Pull request by DanSava**
**Issue**
Initial work on ensemble evaluator implemented using prefect flows

### CODE REVIEW

File: ert_shared/ensemble_evaluator/entity/ensemble.py [hunk 1]
```diff
@@ -245,7 +248,7 @@ def create_step_builder():
 
 
 class _Stage:
-    def __init__(self, id_, steps, status):
+    def __init__(self, id_, steps, status, name=None):
```
Comment (jondequinor) [inline ert_shared/ensemble_evaluator/entity/ensemble.py:251]: I think this should be required. A name is very valuable, even though we will probably generate one for the user.

Lack of names just forces the gui/cli to make something up when visualizing.

File: ert_shared/ensemble_evaluator/entity/prefect_ensamble.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+        # return flow.run(executor=DaskExecutor(address="tcp://192.168.100.6:8786"))
+
+    def evaluate(self, host, port):
+        # executor = DaskExecutor(
+        #     cluster_class="dask_jobqueue.LSFCluster",
+        #     cluster_kwargs=cluster_kwargs,
+        #     debug=True,
+        # )
+        dispatch_url = f"ws://{host}:{port}/dispatch"
+        wait_for_ws(dispatch_url)
```
Comment (jondequinor) [inline ert_shared/ensemble_evaluator/entity/prefect_ensamble.py:279]: I wonder if we could make the `EnsembleEvaluator`guarantee that the ws would be up and running at this point?

File: ert_shared/ensemble_evaluator/entity/prefect_ensamble.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+        return flow.run()
+        # return flow.run(executor=DaskExecutor(address="tcp://192.168.100.6:8786"))
+
+    def evaluate(self, host, port):
+        # executor = DaskExecutor(
+        #     cluster_class="dask_jobqueue.LSFCluster",
+        #     cluster_kwargs=cluster_kwargs,
+        #     debug=True,
+        # )
+        dispatch_url = f"ws://{host}:{port}/dispatch"
```
Comment (jondequinor) [inline ert_shared/ensemble_evaluator/entity/prefect_ensamble.py:278]: we have this in the config already =)

File: ert_shared/ensemble_evaluator/entity/prefect_ensamble.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        for iens in range(self.config['realizations']):
+            stages = []
+            for stage in self.config['stages']:
+                steps = []
+                stage_id = uuid.uuid4()
+                for step in stage["steps"]:
+                    jobs = []
+                    for job in step["jobs"]:
+                        job_id = uuid.uuid4()
+                        jobs.append(_ScriptJob(id_=str(job_id),
```
Comment (jondequinor) [inline ert_shared/ensemble_evaluator/entity/prefect_ensamble.py:178]: Could you create a new type of job? The `ee-legacy` branch turns the `ScriptJob` into a `LegacyJob` that does not have `executable` and `args`.

I thought that, in our discussions, you mentioned that you will never use the `script/unix` type job?

**Review by jondequinor — COMMENTED** [review]
You killed the `SnapshotBuilder`? 😮 

I'm not very sentimental about it, but is your reasoning that it's easier to just create one without it?

File: ert_shared/ensemble_evaluator/entity/prefect_ensamble.py (hunk 4 above)
Comment (mortalisk) [reply ert_shared/ensemble_evaluator/entity/prefect_ensamble.py:178]: Why do we need to distinguish any type of job from another at this point? Why does the evaluator/UI need to know any information about what kind of job it is, or what it is doing?

File: ert_shared/ensemble_evaluator/entity/prefect_ensamble.py (hunk 2 above)
Comment (mortalisk) [reply ert_shared/ensemble_evaluator/entity/prefect_ensamble.py:279]: That sounds reasonable

**Comment by mortalisk** [comment]
We found it to be easier to skip the Builder in stead of making fundamental changes to it's operation to support our stuff. It might change again before we are done. For now it is at least easier to just instantiate directly what we need.

File: ert_shared/ensemble_evaluator/entity/prefect_ensamble.py (hunk 4 above)
Comment (jondequinor) [reply ert_shared/ensemble_evaluator/entity/prefect_ensamble.py:178]: Because we do this today… And we want to keep doing it?

Comment (mortalisk) [reply ert_shared/ensemble_evaluator/entity/prefect_ensamble.py:178]: I think there is something fundamental I am not understanding. What code in the end distinguishes between these types of jobs? And who needs information about what it concretely represents, such as executable?

Comment (jondequinor) [reply ert_shared/ensemble_evaluator/entity/prefect_ensamble.py:178]: In the near future, I assume we will attempt to take ownership—in the `EnsembleEvaluator`—of the task of writing a `jobs.json`. My reasoning is that the legacy evaluator will survive the move to cloud, and that the shared disk assumption in the current code tasked with creating `jobs.json`, will not (should not) work in the cloud.

In order to do this, the evaluator (or the legacy ensemble) needs e.g. `executable`.

Comment (jondequinor) [reply ert_shared/ensemble_evaluator/entity/prefect_ensamble.py:178]: I think you're right that it doesn't really matter whether it's a `LegacyJob` or a `WhateverJob`, as long as they share an interface.

But I feel uneasy making decisions about who needs what when we're supposed to make a general-purpose ensemble evaluator.

**Pull request closed without merging**